### PR TITLE
test: live sync test to check make canonical

### PR DESCRIPTION
**Pull request by fgimenez**
Towards: #10015

Adds a test to verify that the live synced blocks after a backfill sync are eventually made canonical. 

These are the steps of the test 
* creates a chain with 10 more elements than required to trigger a backfill sync
* issues a fcu request to an element of the chain that should trigger a backfill sync
* once done, issues a fcu to the ti of the chain
* verifies that the tip is made canonical.

To make it work I needed to call `make_canonical` in `try_connect_buffered_blocks` on successful insertion, otherwise, this was only done on the code path for downloaded blocks, pls let me know if this is advisable.

There's also additional refactors and improvements in `TestHarness`, like setting up the inserted headers in mock provider to properly simulate a persisted chain.

### CODE REVIEW

File: crates/engine/tree/src/tree/mod.rs [hunk 1]
```diff
@@ -1319,6 +1319,9 @@ where
             match self.insert_block(child) {
                 Ok(res) => {
                     debug!(target: "engine", child =?child_num_hash, ?res, "connected buffered block");
+                    if self.is_sync_target_head(child_num_hash.hash) {
+                        self.make_canonical(child_num_hash.hash);
+                    }
```
Comment (fgimenez) [inline crates/engine/tree/src/tree/mod.rs:1324]: this is the main change in the PR, without this the additional blocks after backfill sync were not being made canonical.

File: crates/engine/tree/src/tree/mod.rs [hunk 2]
```diff
@@ -2152,6 +2167,49 @@ mod tests {
...
+            let mut headers_data: Vec<(B256, Header)> = Vec::with_capacity(blocks.len());
+
+            for block in &blocks {
+                let unsealed_block = block.clone().unseal();
+                block_data.push((block.hash(), unsealed_block.clone().block));
+                headers_data.push((block.hash(), unsealed_block.header.clone()));
+            }
+
+            self.provider.extend_blocks(block_data);
+            self.provider.extend_headers(headers_data);
```
Comment (fgimenez) [inline crates/engine/tree/src/tree/mod.rs:2201]: this allows us to properly query the provider for data persisted.

File: crates/engine/tree/src/tree/mod.rs (hunk 1 above)
Comment (mattsse) [reply crates/engine/tree/src/tree/mod.rs:1324]: oh very nice, yeah I actually noticed this and was wondering

Comment (mattsse) [inline crates/engine/tree/src/tree/mod.rs:1324]: this makes perfect sense, but I think we should add an additional check here that ensures, this is a new block by checking the InsertPayloadOk

https://github.com/paradigmxyz/reth/blob/017372516e1cb7685becd67c5be47f176664d2e5/crates/blockchain-tree-api/src/lib.rs#L256-L257

separately, we should consider not reusing this from legacy tree but roll our own that no longer needs the `BlockAttachment`

**Review by mattsse — CHANGES_REQUESTED** [review]
awesome find, love the tests that make it obvious what scenario is being tested.

I have one suggestion re when we actually call make_canonical on insert

File: crates/engine/tree/src/tree/mod.rs (hunk 1 above)
Comment (fgimenez) [reply crates/engine/tree/src/tree/mod.rs]: > this makes perfect sense, but I think we should add an additional check here that ensures, this is a new block by checking the InsertPayloadOk

makes sense, done here https://github.com/paradigmxyz/reth/pull/10131/commits/b5b1b34770aa6d92821f61fb9cec253469a49994

> separately, we should consider not reusing this from legacy tree but roll our own that no longer needs the `BlockAttachment`

sure I can work on this in a follow-up

**Review by mattsse — APPROVED** [review]

**Pull request merged**